new and improved sql joins demystified (#272)
arcus/education_modules: sql_joins/sql_joins.md

<!--
author:   Joy Payton
email:    paytonk@chop.edu
version:  1.0.0
module_template_version: 3.0.0
language: en
narrator: US English Female
title: SQL Joins
comment: Learn about SQL joins: what they accomplish, and how to write them.
long_description: Usually, data in a SQL database is organized into multiple interrelated tables.  This means you will often have to bring data together from two or more tables into a single dataset to answer your research questions.  This "join" action is accomplished using `JOIN` commands.  This module teaches types of joins, join criteria, and how to write `JOIN` code.
estimated_time: 1 hour

@learning_objectives  
After completion of this module, learners will be able to:

> 

Reviewer: Suggesting on multiple lines doesn't work so well... but I suggest reordering the objectives to match the order of the module:
- Understand the parts of a JOIN
- Describe the "shapes" of SQL JOINs: inner, left, right, and full
- Explain what "join criteria" are

| 11234   | 2021-05-15   | 3    | 3  | 2    | 4    | 12    |
| 86234   | 2021-06-01   | 4    | 4  | 3    | 4    | 15    |
| 32660   | 2021-06-10   | 1    | 1  | 2    | 1    | 5   |
| 86234   | 2022-01-13   | 2    | 2  | 1    | 3    | 8    |
| 41356   | 2022-02-10   | 1    | 3  | 2    | 3    | 10   |

**Table 2: subject\_address**

<!-- data-type="none" class="tight-table" style="font-size:80%"-->
| subj_id  | street_address  | city  | state   | zip  | date_start  | date_end   |
| :--------- | :--------- | :--------- | :--------- | :--------- | :--------- | :--------- |
| 11234   | 123 Main Street   | Smithtown    | PA  | 19000    | 2022-01-01   | `NULL`    |
| 11234   | 123 Oak Lane   | Old Towne    | PA  | 18000   | 2000-01-01    | 2021-12-31    |
| 93452   | 123 Green Blvd  | Kirby    | TN  | 37000    | 2020-05-01    | `NULL`   |

What you want to eventually end up with will be a single table of results that might only have three columns: "subj\_id", "dep\_total", and "zip".  That way you can look at the relationship between depression inventory scale and zip code.

## Forming a JOIN 

There are two basic pieces of information you need to know to write successful joins:

> 

Reviewer: Would it be reasonable to put an entire statement SELECT... FROM... JOIN... ON.. here as an example? It might be helpful to have one "complete" example of how this all fits in before putting ellipses at the start of the code block for the rest of the module, just to remind people what the ellipses are eliding.

Author: agree! changed as the third, fullest example

`ON [table_1_name.field_name] = [table_2_name.field_name]` 

OR

`USING (field_name)` (this method is less frequently used)

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

It can be surprisingly tricky to figure out what makes data "match" or "go together" in your join criteria.  For example, go back to the previous page and see if you can decide what data should be considered "matching" between the "depression\_scale" table and the "subject\_address" table.  Is it as simple as just matching on the subject id ("subj_id")?  Why or why not?

</div>

**Combining Join Type and Join Criteria**
----

You'll combine join type and join criteria by using both the `FROM` component and the `ON` or `USING` component.  Your SQL query might include lines that look something like this: 

```sql
...
FROM table_1 LEFT JOIN table_2
ON table_1.pat_id = table_2.pat_id
...
```

For the rest of this module:

* Let's first talk about join types, 
* Then we'll explain join criteria, and 
* We'll finish up with fuller examples to help you understand how these two elements of a join work together to give you the results you care about.

### Join Types 

> Let's consider the gradebook example we mentioned earlier.  You are assembling grade reports for students.  You have two tables, one called "math\_grades" and one called "language\_grades".  Some students appear in "math\_grades", some in "language\_grades", and some students have rows in both tables.  Depending on your purposes, you might want any one of several types of joins, each with its own SQL keyword combination.  Let's imagine a Venn diagram, as shown below:

Author: ```suggestion
Let's consider the gradebook example we mentioned earlier.  You are assembling grade reports for students.  You have two tables, one called "math\_grades" and one called "language\_grades".  Some students appear in "math\_grades", some in "language\_grades", and some students have rows in both tables.  Depending on your purposes, you might want any one of several types of joins, each with its own SQL keyword combination.  Let's imagine a Venn diagram of side by side circles, the one on the left representing the group of students who appear in the math\_grades table, and the circle on the right representing the group of students who appear in the language\_grades table.  There's some overlap of these two circles, which consists of the students who appear in both tables.
```

<lia-keep>

> 

Reviewer: Do you know how a screen reader will handle this infographic? I'm not sure how or if the text in it becomes alt text.

  vertical-align: middle;
}

#c1 {
  top: 0;
  left: 0;
  background-color: rgba(255,0,0,.25);
}

#c2 {
  top: 0;
  left: 12em;
  background-color: rgba(0,255,0,.25);
}

#c1 p{
  padding-right: 2em;
}
#c2 p {
    padding-left: 12em;
    width: 22em;
}

#j12 {
    position: absolute;
    left: 12em;
    width: 11em;
    height: 22em;
    padding: 1em;
    text-align: center;
}

#j12 p {
  height: 22em;
  width: 10em;
  display: table-cell;
  vertical-align: middle;
}
</style>
<div id="infographic">
  <div id="c1" class="circle">
    <p>Left Side: <br/>Only Math Grades</p>
  </div>

  <div id="c2" class="circle">
    <p>Right Side: <br/>Only Language Grades</p>
  </div>

  <div id="j12" class="joined">
    <p>
      Overlap: <br/>Both Math and Language Grades
    </p>
  </div>
</div>

</lia-keep>

There are 4 basic join types that can be used.  We'll go into more detail about how each one works in the next few pages.  

Here, we simply provide a visual highlight to indicate the part of the data that's considered for inclusion with each type of join.

* **`INNER JOIN`**.  ![Overlapping circles with only the inner, overlapping part highlighted](media/inner.png) 

> * **`LEFT JOIN`**.  ![Overlapping circles with the left circle, including the overlapping part of the left circle, highlighted](media/left_outer.png) 

Reviewer: ```suggestion
* **`LEFT JOIN`**.  ![Overlapping side-by-side circles with the left circle, including the overlapping part of the left circle, highlighted](media/left_outer.png) 
```

> * **`RIGHT JOIN`**.  ![Overlapping circles with the right circle, including the overlapping part of the right circle, highlighted](media/right_outer.png) 

Reviewer: ```suggestion
* **`RIGHT JOIN`**.  ![Overlapping side-by-side circles with the right circle, including the overlapping part of the right circle, highlighted](media/right_outer.png) 
```

* **`FULL JOIN`**.  ![Overlapping circles with both circles highlighted](media/full_outer.png) 

#### `INNER JOIN`

Here, let's consider the left table to be **math\_grades** and the right table to be **language\_grades**.

![Overlapping circles with only the inner, overlapping part highlighted](media/inner.png) 

**What is it?**
-----

An `INNER JOIN` (and this is the default behavior of `JOIN` without any modifying words) finds matching grade data falling in the "Both Math and Language Grades" overlap section.  This is often the data you want to capture.  

**Why use it?**
-----

If you're conducting research on the correlation between math and language grades, this is the join you want.  If a student lacks one or the other grade, their data isn't useful to you, so you don't want it. 

**Results**
-----
The result of an inner join will be a table that only has rows of data for students who appear in both tables.  If a student is missing in one or the other table of grades, that student won't appear in your result set.

**Code**
----

An `INNER JOIN` shows up in code like this:

```sql
...

> FROM math_grades INNER JOIN language_grades

Reviewer: ```suggestion
FROM math_grades INNER JOIN language_grades
...
```
Is it possible to join tables without an ON or a USING statement? If so, the second ellipses might not be necessary, but I think there are.

```

or

```sql
...

> FROM math_grades JOIN language_grades

Reviewer: ```suggestion
FROM math_grades JOIN language_grades
...
```

```

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

Note that the word `JOIN` by itself means `INNER JOIN`.  This is because this is the most frequently used kind of join.  Make sure you remember that the default kind of join, without any keywords, is an inner join, which may or may not be what you want!
</div>

#### `LEFT JOIN`

Here, let's consider the left table to be **math\_grades** and the right table to be **language\_grades**.

> ![Overlapping circles with the left circle, including the overlapping part of the left circle, highlighted](media/left_outer.png) 

Reviewer: ```suggestion
![Overlapping side-by-side circles with the left circle, including the overlapping part of the left circle, highlighted](media/left_outer.png) 
```

**Why use it?**
-----
Maybe you want this data because as the chair of the mathematics department, you want to see what students are strongest, in math, and also in language if known, in order to select who to award a math prize to.  Perhaps you'll use language grades as a tie-breaker, for example?

**Results**
-----
The result of a left join will be a table that has:

* Rows of data for students who appear in the left table only
* Rows of data for students who appear in both tables. 

When there's no matching data from the right table to join to the students who only appear in the left table, `NULL` values (empty cells) are added to the appropriate column names.

**Code**
----

A `LEFT JOIN` shows up in code like this.  Note that one table is listed first, and is therefore on the left, and the other table is listed second, and is therefore on the right (because English is written left to right).

```sql
...

> FROM math_grades LEFT JOIN language_grades

Reviewer: ```suggestion
FROM math_grades LEFT JOIN language_grades
...
```

```

or

```sql
...
FROM math_grades LEFT OUTER JOIN language_grades
```

#### `RIGHT JOIN`

Here, let's consider the left table to be **math\_grades** and the right table to be **language\_grades**.

> ![Overlapping circles with the right circle, including the overlapping part of the right circle, highlighted](media/right_outer.png) 

Reviewer: ```suggestion
![Overlapping side-by-side circles with the right circle, including the overlapping part of the right circle, highlighted](media/right_outer.png) 
```

**What is it?**
-----

A `RIGHT JOIN` (or a `RIGHT OUTER JOIN`) finds unmatched data in "Only Language Grades" -- that's the "outer" part on the right -- and also finds matching data in "Both Math and Language Grades" -- the overlap section.  It would exclude student data in the "Only Math Grades" section (which is on the left).  

**Why use it?**
-----

Maybe you want this data to create a graph of language grade distribution, and want to enrich it with some statistics about math grades where available, but it doesn't matter if some students lack math grades. 

**Results**
-----

The result of a right join will be a table that has:

* Rows of data for students who appear in the right table only
* Rows of data for students who appear in both tables. 

When there's no matching data from the left table to join to the students who only appear in the right table, `NULL` values (empty cells) are added to the appropriate column names.

**Code**
----

A `RIGHT JOIN` shows up in code like this.  Recall that "left" and "right" refer to the order in which the table names are typed in the `FROM` statement.

```sql
...

> FROM math_grades RIGHT JOIN language_grades

Reviewer: ```suggestion
FROM math_grades RIGHT JOIN language_grades
...
```

```

or

```sql
...

> FROM math_grades RIGHT OUTER JOIN language_grades

Reviewer: ```suggestion
FROM math_grades RIGHT OUTER JOIN language_grades
...
```

```

> <div class = "important">

Author: ```suggestion
<div class = "options">
```

> <b style="color: rgb(var(--color-highlight));">Important note</b><br>

Reviewer: ```suggestion
<b style="color: rgb(var(--color-highlight));">Another option</b><br>
```

<code>
FROM math\_grades RIGHT JOIN language\_grades
</code>
<br/><br/>
Can be rewritten as: 
<br/><br/>
<code>
FROM language\_grades LEFT JOIN math\_grades
</code>

In case you're wondering why you only see left joins, and never right joins, in other people's code: it's simply a convention some people follow, to swap table name order to always have left joins.
</div>

#### `FULL JOIN`

Here, let's consider the left table to be **math\_grades** and the right table to be **language\_grades**.

 ![Overlapping circles with both circles highlighted](media/full_outer.png) 

**What is it?**
-----

A `FULL JOIN` (or sometimes you'll see `FULL OUTER JOIN`) will consider all the data -- all of the outer, unmatched data on the right (language\_grades) and left (math\_grades) tables, as well as the inner overlapping data which has matching data from both tables.  

**Why use it?**
-----

This kind of join is important if you want to create a grade book that shows student grades for each student, including their math grades and/or their language grades. 

**Results**
-----

The result of a right join will be a table that has:

> * Rows of data for students who appear in the left table only
> * Rows of data for students who appear in the right table only
> * Rows of data for students who appear in both tables. 

Reviewer: ```suggestion
* Rows of data for students who appear in the left table only
* Rows of data for students who appear in both tables
* Rows of data for students who appear in the right table only.

```

Reviewer: This is definitely a peek into my brain, feel free to ignore entirely.

When there's no matching data from the one of the tables to join to the data you included from the other table, `NULL` values (empty cells) are added.

**Code**
----

A `FULL JOIN` shows up in code like this:

```sql
...

> FROM math_grades FULL JOIN language_grades

Reviewer: ```suggestion
FROM math_grades FULL JOIN language_grades
...
```

```

or

```sql
...

> FROM math_grades FULL OUTER JOIN language_grades

Reviewer: ```suggestion
FROM math_grades FULL OUTER JOIN language_grades
...
```

#### Equality, Example 1

Do you have subject identifiers or student ID numbers in two different tables?  This shared information can be used to connect (join) data from these tables, based on the identifier being equal.  

For example, if the subject ID matches, a row from table A and a row from table B will be joined.  If the subject ID doesn't match, these rows won't be joined.  Maybe we're trying to match lung cancer occurrence and smoking exposure in the same row:

**disease**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | lung\_cancer |
| :--------- | :--------- | 
| 3  | TRUE |
| 5 | FALSE  |
| 8  | FALSE |

**smoking**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | smoking\_pack\_years |
| :--------- | :--------- | 
| 2  | 10 |
| 3 | 10  |
| 4  | 0 |

In code, we usually use an **ON** statement:

```sql
...
ON disease.subject_id = smoking.subject_id
```

> In this case, we're comparing the equality of two fields that **have the same name**, so we could also use **USING**.  This special word only applies when you're looking for a perfect match between fields that have matching names, too.  It's okay if you never use `USING` and prefer to always stick with `ON`, which is more multi-purpose.
> 
> ```sql
> ...
> USING(subject_id)
> ```

Reviewer: ```suggestion
<div class = "options">
<b style="color: rgb(var(--color-highlight));">Another option</b><br>

In this case, we're comparing the equality of two fields that **have the same name**, so we could also use **USING**.  This special word only applies when you're looking for a perfect match between fields that have matching names, too.  It's okay if you never use `USING` and prefer to always stick with `ON`, which is more multi-purpose.

```sql
...
USING(subject_id)
```

</div>
```

With either of the above code snippets:

> * subject 3 from disease matches with subject 3 from smoking <i class="fa fa-check" aria-hidden="true"></i>

Reviewer: Does the `aria-hidden="true"` tell screenreader what to say? That is cool!

Author: No, I think it means "hide this graphic" but honestly I just copy-pasted from the FA website :)

Reviewer: in that case, I think all of these fun check/X marks should be accompanied or replaced by a word like "included" or "excluded" because the symbol is the only indication of which are included/excluded. 

See this conversation about how these need alt text: https://stackoverflow.com/questions/29393575/font-awesome-support-for-screen-readers-and-accessibility

Author: I think the best thing is just to delete... these are really decorations that don't add much, really.

This also shows that figuring out your join criteria requires close attention: we probably don't mean to match student 11's math grades with student 14's language grades!

</div>

#### Non-Equality and More

Sometimes you don't need equality as your condition.  For example, in our example from our multi-site mental health research (see the first page of this module for a reminder), let's say we want to associate a particular depression score with a particular address only if the depression inventory was given between the start and end dates of residency at that address. In a case like that, you might see something like:

```sql
...
ON depression_scale.date BETWEEN 
   subject_address.date_start AND 
   subject_address.date_end
```

**Multiple conditions**

Of course, in the example of the depression inventory, we'd also want to make sure that there was a match on subject identifier (you don't want to match Lakshmi's depression score with Larry's address, just because the dates worked!).  You can combine conditions too.  Here, we look for an exact match on subject ID and a date match that's between the correct dates:

```sql

> ON depression_scale.subj_id = subject_address.subj_id AND

Reviewer: ```suggestion
...
ON depression_scale.subj_id = subject_address.subj_id AND
```

   depression.date BETWEEN 
      subject_address.date_start AND 
      subject_address.date_end
```

#### Getting Really Complicated

Let's keep thinking about our depression inventories and our goal of matching depression scores to addresses in our study. What if some addresses don't have end dates?  This could be because the subject is currently still living there.  There are some addresses with `NULL` end dates in our data, so this isn't an academic question.

You can create arbitrarily complex **boolean logic** (or **boolean algebra**), using AND, OR, NOT, and parentheses as needed.  Much as in math, there's an order of operations in this kind of logic, and you might need several sets of parentheses to make sure you're applying the conditions correctly.  For example, see below.  We've added comments to help illustrate the logic.

```sql

> ON depression.subj_id = subject_address.subj_id AND 

Reviewer: ```suggestion
...
ON depression.subj_id = subject_address.subj_id AND 
```

* An `INNER JOIN` involving equality matching on three fields: two identifiers and one date field
* A `RIGHT JOIN` involving matching data with a "less than" relationship, such as a 3 month old weight measurement being less than a 1 month old weight measurement
* and many more!

Instead of being exhaustive, we've concentrated on the most frequent use case you'll encounter over and over: equality.  We'll go over each join type (inner, left, right, and full) on a simple equality matching a single field from each table.

#### `INNER JOIN` and Equality Condition

To understand what an `INNER JOIN` with equality looks and acts like practically, let's go to a simple example of two tables we used earlier in this module:

**disease**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | lung\_cancer |
| :--------- | :--------- | 
| 3  | TRUE |
| 5 | FALSE  |
| 8  | FALSE |

**smoking**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | smoking\_pack\_years |
| :--------- | :--------- | 
| 2  | 10 |
| 3 | 10  |
| 4  | 0 |

> And let's combine our join criteria (subject\_id matching) with our join type (inner).  We won't show the whole query here, just the pertinent parts that describe the join:

Reviewer: Would you consider having the first line of the code be `SELECT *` instead of `...`?
Given that the entirety of the code seems no longer than the abbreviated code, and including `SELECT *` doesn't seem that distracting. I'm personally more distracted that it isn't there.

```sql
...
FROM disease JOIN smoking
ON disease.subject_id = smoking.subject_id
```

This is the result we would get.  It only includes data for subjects appearing in both tables:

<!-- data-type="none" class="tight-table"-->
| subject\_id  | lung\_cancer | smoking\_pack\_years |
| :--------- | :--------- | :--------- | 
| 3  | TRUE | 10 |

> <div class = "important">
> <b style="color: rgb(var(--color-highlight));">Important note</b><br>

Reviewer: ```suggestion
<div class = "options">
<b style="color: rgb(var(--color-highlight));">Another option</b><br>
```

**disease**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | lung\_cancer |
| :--------- | :--------- | 
| 3  | TRUE |
| 5 | FALSE  |
| 8  | FALSE |

**smoking**

<!-- data-type="none" class="tight-table"-->
| subject\_id  | smoking\_pack\_years |
| :--------- | :--------- | 
| 2  | 10 |
| 3 | 10  |
| 4  | 0 |

And let's combine our join criteria (subject\_id matching) with our join type (full).  We won't show the whole query here, just the pertinent parts that describe the join:

```sql
...
FROM disease FULL JOIN smoking
ON disease.subject_id = smoking.subject_id
```

This is the result we would get.  Each subject is represented here, both the ones who appear in the left table and in the right table.

<!-- data-type="none" class="tight-table"-->
| subject\_id  | lung\_cancer | smoking\_pack\_years |
| :--------- | :--------- | :--------- | 
| 3  | TRUE | 10 |
| 5 | FALSE  | `NULL` |
| 8  | FALSE | `NULL` |
| 2  | `NULL` | 10 |
| 4  | `NULL` | 0 |

When there's no matching data from the one of the tables to join to the data you included from the other table, `NULL` values (empty cells) are added.

## Additional Resources

> * Want to understand the basics of boolean algebra?  Check out a good tutorial at https://ryanstutorials.net/boolean-algebra-tutorial/boolean-algebra.php.

Reviewer: ```suggestion
* Want to understand the basics of boolean algebra?  Check out a good tutorial at https://ryanstutorials.net/boolean-algebra-tutorial/boolean-algebra.php
```
